File: lib/121.best_time_to_buy_and_sell_stock.dart
import 'dart:math';

void main() {
  print(maxProfit([7, 1, 5, 3, 6, 4]));
  print(maxProfit([7, 6, 4, 3, 1]));
  print(maxProfit([1, 2]));
}

int maxProfit(List<int> prices) {
  var maxPrice = 0;
  var minValue = prices.first;

  for (int i = 1; i < prices.length; i++) {
    maxPrice = max(maxPrice, prices[i] - minValue);
    minValue = min(minValue, prices[i]);
  }

  return maxPrice;
}

File: src/121.best_time_to_buy_and_sell_stock.kt
fun main() {
    println(maxProfit(intArrayOf(7, 1, 5, 3, 6, 4)))
    println(maxProfit(intArrayOf(7, 6, 4, 3, 1)))
    println(maxProfit(intArrayOf(1, 2)))
}

fun maxProfit(prices: IntArray): Int {
    var maxPrice = 0
    var minValue = prices.first()

    for (i in 1..<prices.size) {
        maxPrice = Math.max(maxPrice, prices[i] - minValue)
        minValue = Math.min(minValue, prices[i])
    }

    return maxPrice
}